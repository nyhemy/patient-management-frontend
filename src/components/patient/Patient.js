import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import styles from './Patient.module.css';
import loadImg from '../ajax-loader.gif';

const axios = require('axios').default;

/**
 * Component which takes data from backend Patient entity and displays it
 *
 * @param {*} props are props passed to component when it is called
 */
const Patient = (props) => {
  const {
    id, firstName, lastName, age, gender
  } = props;

  const history = useHistory();
  const [loading, setLoading] = useState(false);
  const [errorMsg, setErrorMsg] = useState('');

  /**
   * Redirects to PatientDetails
   */
  const patientDetails = () => {
    history.push(`/patients/${id}`);
  };

  /**
   * Handles delete request for Patient to API
   */
  const deletePatient = () => {
    setLoading(true);
    axios.delete(`http://localhost:8080/patients/${id}`, {
      headers: {
        'Content-Type': 'application/json',
        mode: 'cors',
        Authorization: `Bearer ${sessionStorage.getItem('token')}`
      },
      data: {
        id
      }
    })
      // eslint-disable-next-line no-unused-vars
      .then((response) => {
        setLoading(false);
        window.location.reload();
      })
      // eslint-disable-next-line no-unused-vars
      .catch((error) => {
        setLoading(false);
        if (error.response) {
          switch (error.response.status) {
            case 409:
              setErrorMsg('Cannot delete patient with existing encounters');
              break;
            default:
              setErrorMsg('Oops something went wrong');
              break;
          }
        } else {
          setErrorMsg('Oops something went wrong');
        }
      });
  };

  return (
    <div className={styles.card}>
      {loading
        ? <img src={loadImg} alt="loading..." />
        : (
          <div>
            <div>
              Name:
              {` ${firstName} ${lastName}`}
            </div>
            <div>
              Age:
              {` ${age}`}
            </div>
            <div>
              Gender:
              {` ${gender}`}
            </div>
            <div>
              <button type="button" onClick={patientDetails}>Select</button>
              <button type="button" onClick={deletePatient}>Delete</button>
            </div>
          </div>
        )}
      {errorMsg !== '' && <h2 className={styles.notification}>{errorMsg}</h2>}
    </div>
  );
};

export default Patient;
