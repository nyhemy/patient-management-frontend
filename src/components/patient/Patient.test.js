import React from 'react';
import { cleanup, render, screen, act, getByText, waitFor, getByAltText, queryByAltText, fireEvent } from '@testing-library/react';
import { unmountComponentAtNode } from 'react-dom';
import renderer from 'react-test-renderer';
import fetchMock from 'jest-fetch-mock';
import { MemoryRouter } from 'react-router-dom';
import Patient from './Patient';
// import Patients from '../patients/Patients';
import '@testing-library/jest-dom/extend-expect';

fetchMock.enableMocks();

beforeEach(() => {
  fetch.resetMocks();
});

afterEach(cleanup);

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

it('renders w/o crashing and matches snapshot', () => {
  const div = document.createElement('div');
  render(<Patient />, div);

  const component = renderer.create(<Patient />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('shows error for 409', async () => {
  fetch.mockResponseOnce('{ "id": 1 }', { status: 409, headers: { 'content-type': 'application/json' } });
  const { getByText, queryByText } = render(<Patient />);

  const deleteButton = getByText('Delete');
  fireEvent.click(deleteButton);
  await waitFor(() => getByText('Cannot delete patient with existing encounters'));
  expect(screen.queryByText('Cannot delete patient with existing encounters')).toBeTruthy();
});

it('shows generic error', async () => {
  fetch.mockResponseOnce('{ "id": 1 }', { status: 500, headers: { 'content-type': 'application/json' } });
  const { getByText, queryByText } = render(<Patient />);

  const deleteButton = getByText('Delete');
  fireEvent.click(deleteButton);
  // await waitFor(() => expect(screen.queryByText('Oops something went wrong')).toBeTruthy());
  await waitFor(() => getByText('Oops something went wrong'));
  expect(screen.queryByText('Oops something went wrong')).toBeTruthy();
});

// it('shows catch error', async () => {
//   fetch.mockResponseOnce('{ "id": 1 }', { status: 500, headers: { 'content-type': 'giberish' } });
//   const { getByText, queryByText } = render(<Patient />);

//   const deleteButton = getByText('Delete');
//   fireEvent.click(deleteButton);
//   // await waitFor(() => expect(screen.queryByText('Oops something went wrong')).toBeTruthy());
//   await waitFor(() => getByText('Oops something went haywire'));
//   expect(screen.queryByText('Oops something went haywire')).toBeTruthy();
// });

it('deletes', () => {
  const { getByText, queryByText } = render(<Patient />);

  expect(screen.queryByText('Name:', { exact: false })).toBeTruthy();
  const deleteButton = getByText('Delete');
  fireEvent.click(deleteButton);
  expect(screen.queryByText('Name:', { exact: false })).toBeFalsy();
  expect(screen.queryByText('Oops something went wrong')).toBeFalsy();
});

it('redirects to patient details', () => {
  const { getByText, queryByText } = render(<Patient />);

  const selectButton = getByText('Select');
  fireEvent.click(selectButton);
  expect(mockHistoryPush).toHaveBeenCalledWith('/patients/undefined');
});

// const { getByText, queryByText } = render(<Patient />);

//   fireEvent.click(getByText('Select'));
//   expect(screen.queryByText('Name:', { exact: false })).toBeFalsy();
//   expect(mockHistoryPush).toHaveBeenCalledWith('http://localhost:8080/patients/null');

it('shows data and buttons', async () => {
  // fetch.mockResponseOnce(JSON.stringify([
  //   {
  //     id: 1,
  //     firstName: "Test",
  //     lastName: "Last",
  //     email: "tl@gmail.com",
  //     street: "Yo Dr",
  //     city: "Heyvalley",
  //     state: "NH",
  //     postal: "00000",
  //     age: 23,
  //     height: 64,
  //     weight: 112,
  //     insurance: "Wooshoo Inc",
  //     gender: "female",
  //     ssn: "000-00-0000"
  //   }
  // ]));
  // const { getByText, queryByText } = render(<Patients />);
  // await waitFor(() => getByText('Name: Test Last'));
  // expect(screen.queryByText('Name: Test Last')).toBeTruthy();
  // expect(screen.queryByText('Age: 23')).toBeTruthy();
  // expect(screen.queryByText('Gender: female')).toBeTruthy();
  // expect(screen.queryByText('Select')).toBeTruthy();
  // expect(screen.queryByText('Delete')).toBeTruthy();
  const { getByText, queryByText } = render(<Patient />);
  await waitFor(() => getByText('Name: undefined undefined'));
  expect(screen.queryByText('Name: undefined undefined')).toBeTruthy();
  expect(screen.queryByText('Age: undefined', { exact: false })).toBeTruthy();
  expect(screen.queryByText('Gender: undefined', { exact: false })).toBeTruthy();
  expect(screen.queryByText('Select')).toBeTruthy();
  expect(screen.queryByText('Delete')).toBeTruthy();
});
