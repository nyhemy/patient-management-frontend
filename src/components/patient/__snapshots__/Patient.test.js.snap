// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app component matches snapshot 1`] = `
<div
  className="card"
>
  <div>
    <div>
      Name:
       undefined undefined
    </div>
    <div>
      Age:
       undefined
    </div>
    <div>
      Gender:
       undefined
    </div>
    <div>
      <button
        onClick={[Function]}
        type="button"
      >
        Select
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;
